repaired the ability of nano to be able to output DEBUG messages

glynnbird opened the pull request:
I noticed that when I ran my application with

```
DEBUG="*" node myapp.js
```

I wasn't seeing any debug output from Nano. I had a look at the code and there seemed to be some confusion about when Nano uses the `debug` package and where in the code a custom `log` function can be provided.

I have refactored the logger.js code so that it represents the default `debug` package function. The code that optionally switches in a custom, user-supplied log function was already handled in nano.js but there seemed to be some of this code leftover in logger.js, which caused no debug messages to ever work (unless you provided your own function).

With this PR, the debug messages now work with
- DEBUG="*" node mycode.js
- or export DEBUG=nano

I've also added a paragraph in README.md to explain this to users.

glynnbird committed:
repaired the ability of nano to be able to log its messages when a user sets the DEBUG environment variable

dscape commented on lib/logger.js:
```diff
@@ -11,18 +11,12 @@
+module.exports = function logging() {
+  return function log() {
+    var eventId = prefix + (~~(Math.random() * 1e9)).toString(36);
+    debug.call(this, eventId, [].slice.call(arguments, 0));
```
thanks for the pull request. however i'm confused by this. 

the log function is normally used to log, and during development can be overridden by debug. seems like this just makes it be a debug message all the time no?

dscape commented on README.md:
```diff
@@ -804,6 +804,32 @@ alice.attachment.get('rabbit', 'picture.png').pipe(fs.createWriteStream('/tmp/ra
 
 then open `/tmp/rabbit.png` and you will see the rabbit picture.
 
+### debugging
```
docs. always great :+1:

dscape commented on tests/integration/shared/log.js:
```diff
@@ -19,12 +19,7 @@ var harness = helpers.harness(__filename);
-  log('testing 1234');
+  var log = logger();
+  assert.equal(typeof log, 'function');
+  assert.end();
```
once again, seems like you removed api stable behavior and replaced it with what was convenient for your use case?

while you might not like the behavior, many people might be relying on it in production and we can't do breaking changes like this

if you have a rationale for it please explain, maybe a breaking version can be released then

glynnbird commented:
The code that allows a user defined function to be supplied  is here:

* https://github.com/dscape/nano/blob/master/lib/nano.js#L45

This remains in place. The problem was that this code was *also* repeated in logger.js. If no overriding log function was provided, then *no logging actually happened*!

The modified test reflects the behaviour that logger module should return a function. Sorry if this wasn't clear in the PR.

jhs commented:
Hi, @dscape.

I think the key point here, with which we can all agree, is that this API is not specified concretely nor tested against. AFAICT, the test suite does not test logging at all. (Yes, it hooks the debug package into the logging, but it does not actually test or confirm things.)

So, I propose we create a new PR which tests the logger API (in effect defining it concretely), and then after that, we can continue this conversation?

jo commented:
:+1:

jhs commented:
I think the confusion is this: if you look at logger.js, it exports a function which returns a function which returns a function. (That is not an composition error. A function returns a function which returns a third function.)

In other words, by looking at this code, one would expect to see it used like so:

    var loggingSystem = require("./logger.js");

    function actuallyLog(eventId, args) {
      console.log(eventId + " " + JSON.stringify(args)); // Simple example.
    }

    var logConfig = {"log": actuallyLog};
    var loggerFactory = loggingSystem(logConfig);

    // Suppose we have two concerns ("prefixes"): a generic logger and then one about, say, bats.
    var genLogger = loggerFactory();
    var batLogger = loggerFactory("bats");

    genLogger("Begin!");
    genLogger("Something happened", {count:3});

    batLogger("Strike!", {type:"aluminum"});
    batLogger("Home run!", {type:"wood"});
    batLogger("Found a mosquito", [0.3, 0.2, 0.1], {method:"echolocation"});

    genLogger("Finished");

Output:

    13qqx5 ["Begin!"]
    13qqx5 ["Something happened",{"count":3}]
    bats-8w2wes ["Strike!",{"type":"aluminum"}]
    bats-8w2wes ["Home run!",{"type":"wood"}]
    bats-8w2wes ["Found a mosquito",[0.3,0.2,0.1],{"method":"echolocation"}]
    13qqx5 ["Finished"]

In other words, we would expect three function calls deep, through the metafunctions, to actually log events.

If you look at the code in nano.js: https://github.com/dscape/nano/blob/v6.1.4/lib/nano.js#L45

    var log = typeof cfg.log === 'function' ? cfg.log : logger(cfg);

Here I am confused. The logging module will check `cfg.log` and react accordingly: https://github.com/dscape/nano/blob/v6.1.4/lib/logger.js#L18-L19

    var log = cfg && cfg.log;
    var logStrategy = typeof log === 'function' ? log : debug;

 But the code in `nano.js` seems to guarantee that `logger()` will never be called with a `.log` value. In other words, there seems to be no way I can pass my logging function *into* logger.js, because `nano.js` will short-circuit that process.

My next confusion is, that the function `log()` in nano.js, is called with bits of information worth logging: the request (line 179), or errors (lines 191, 209, and 213). But, I would expect `log(foo)` not to actually log something, but rather to *return a logger function*. In other words, "log" from nano.js:45 references "logEvent" from logger.js:28.

If that is the case, then if somebody provides their own `cfg.log` function, fine, they can do what they want. However if they wish to rely on the default Debug package behavior, there is actually no way to do that. In other words, `env DEBUG="nano/logger" nano.js` would never actually output any messages.

Perhaps I have totally overthought myself into misunderstanding. It wouldn't be the first time. But if not, maybe you can agree that perhaps it is time that we nail down the logging API. If anybody has any feedback at all, then I would be grateful and I will happily write up some unit tests to codify that. Thank you very much!

revington committed:
Test with 4.2 (LTS) and 5.5 (Stable)

revington committed:
removed the osx target since it is not supported by travis

briandela committed:
Update README.md

Change `cloudant.use` to `nano.use`.

satazor committed:
Add db.info doc.

satazor committed:
Ensure fetch params is optional, fixes #319.

cttttt committed:
ct: Test for optional body in db.atomic

cttttt committed:
ct: Allow body to be optional in db.atomic.

cttttt committed:
ct: Fix mocks.

KangTheTerrible committed:
Validate docName parameter to avoid db delete

pmanijak committed:
Update express.js example for nano 6 and express 4

There were a few places where this file was out of date.

Express 4:
* express.createServer() has been deprecated
* passing the status via send() has been deprecated
* convention is to use 'req' and 'res' instead of 
  'request' and 'response'  

Nano 6:
* error['status-code'] is now error.statusCode

Also there was a syntax error (too many curly braces).

glynnbird committed:
version bump for dependencies

glynnbird committed:
fix mocked tests

glynnbird committed:
modified tests to work with latest dependencies and CouchDB 2.0

glynnbird committed:
modify tests to work mocked and unmocked with CouchDB 2.0 from Docker

glynnbird committed:
ensure CouchDB 2.0 Docker is stopped after each test run

glynnbird committed:
Nano improvements

The improvements:
* Node 0.12 is no longer supported
* Only JSON encode parameters which are not strings to avoid double encoding as per https://github.com/cloudant/nodejs-cloudant/issues/89
* Update package.json for this repo's project name
* Added GET /_uuids api call
* Added selective encoding of string parameters to the search function

dominicbarnes committed:
add reference to server in doc scope

carlosduclos committed:
Adding replication using the "_replicator" database

Starting with 1.2.0 CouchDB added a new system database called
"_replicator" to handle replications jobs.

Replications are now created as entries on that database and
the server will schedule and perform the replication
accordingly. Entries in the "_replicator" db will be updated.

This means that replication now is a completely asynchronous job
that is not guaranteed to run right after the replication was started.

This commit adds three new object with three object to handle this new
type of replication:

- replication.enable: To enable the replication of a database.
- replication.query: To query the status of a replication job.
- replication.disable: To disable the replication of a database.

More information on this type of replication can be found:
- https://wiki.apache.org/couchdb/Replication#from_1.2.0_onward
- http://guide.couchdb.org/draft/replication.html
- https://gist.github.com/fdmanana/832610

[Addendum after merging with the new repo]
Fixing tests for uuids, since they were not passing.

idearat committed:
Ensure 'qs' is real whenever qs will be manipulated.

carlosduclos commented:
Hi,
The NanoJS repository has been merged into apache/couchdb-nano. Could you close this pull request and open it there instead?
Thanks in advance

glynnbird committed:
Merge remote-tracking branch 'upstream/master'
